roachtest: cdc/tpcc-1000/rangefeed=true failed

[(roachtest).cdc/tpcc-1000/rangefeed=true failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1823985&tab=buildLog) on [master@055561809b95488bff2cad19422e7f4a7472e3a2](https://github.com/cockroachdb/cockroach/commits/055561809b95488bff2cad19422e7f4a7472e3a2):

```
		       56.0s     1289            4.0           11.4  22548.6  33286.0  33286.0  33286.0 stockLevel
		    _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		       57.0s     1371            3.0           11.9  23622.3  24696.1  24696.1  24696.1 delivery
		       57.0s     1371            0.0           76.9      0.0      0.0      0.0      0.0 newOrder
		       57.0s     1371            3.0           11.2  22548.6  23622.3  23622.3  23622.3 orderStatus
		       57.0s     1371            0.0           97.1      0.0      0.0      0.0      0.0 payment
		       57.0s     1371            1.0           11.2  25769.8  25769.8  25769.8  25769.8 stockLevel
		       58.0s     1581            6.0           11.8  23622.3  24696.1  24696.1  24696.1 delivery
		       58.0s     1581            0.0           75.6      0.0      0.0      0.0      0.0 newOrder
		       58.0s     1581           11.0           11.2  24696.1  26843.5  26843.5  26843.5 orderStatus
		       58.0s     1581            1.0           95.4  33286.0  33286.0  33286.0  33286.0 payment
		       58.0s     1581            4.0           11.1  22548.6  23622.3  23622.3  23622.3 stockLevel
		       59.0s     2142            7.0           11.7  26843.5  28991.0  28991.0  28991.0 delivery
		       59.0s     2142            0.0           74.3      0.0      0.0      0.0      0.0 newOrder
		       59.0s     2142            4.0           11.1  25769.8  30064.8  30064.8  30064.8 orderStatus
		       59.0s     2142            1.0           93.8  30064.8  30064.8  30064.8  30064.8 payment
		       59.0s     2142            7.0           11.0  23622.3  26843.5  26843.5  26843.5 stockLevel
		       60.0s     2223           10.0           11.6  26843.5  30064.8  30064.8  30064.8 delivery
		       60.0s     2223            0.0           73.1      0.0      0.0      0.0      0.0 newOrder
		       60.0s     2223            4.0           11.0  25769.8  26843.5  26843.5  26843.5 orderStatus
		       60.0s     2223            2.0           92.3  31138.5  36507.2  36507.2  36507.2 payment
		       60.0s     2223            2.0           10.8  23622.3  26843.5  26843.5  26843.5 stockLevel
		    _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		       61.0s     2379            2.0           11.5  25769.8  25769.8  25769.8  25769.8 delivery
		       61.0s     2379            0.0           71.9      0.0      0.0      0.0      0.0 newOrder
		       61.0s     2379            0.0           10.8      0.0      0.0      0.0      0.0 orderStatus
		       61.0s     2379            0.0           90.8      0.0      0.0      0.0      0.0 payment
		       61.0s     2379            4.0           10.7  25769.8  26843.5  26843.5  26843.5 stockLevel
		       62.0s     2663            3.0           11.4  28991.0  31138.5  31138.5  31138.5 delivery
		       62.0s     2663            0.0           70.7      0.0      0.0      0.0      0.0 newOrder
		       62.0s     2663            0.0           10.6      0.0      0.0      0.0      0.0 orderStatus
		       62.0s     2663            0.0           89.3      0.0      0.0      0.0      0.0 payment
		       62.0s     2663            4.0           10.6  25769.8  27917.3  27917.3  27917.3 stockLevel:
		  - context canceled

	cluster.go:2368,cdc.go:224,cdc.go:521,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.cdcBasicTest
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:224
		    main.registerCDC.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:521
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 3: dead

```

<details><summary>More</summary><p>


Artifacts: [/cdc/tpcc-1000/rangefeed=true](https://teamcity.cockroachdb.com/viewLog.html?buildId=1823985&tab=artifacts#/cdc/tpcc-1000/rangefeed=true)
Related:
- #46355 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202003200044_v20.1.0-beta.3](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202003200044_v20.1.0-beta.3) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #46137 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202003112118_v20.1.0-beta.3](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202003112118_v20.1.0-beta.3) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #45437 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44971 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44572 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202001302015_v19.2.3](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202001302015_v19.2.3) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Acdc%2Ftpcc-1000%2Frangefeed%3Dtrue.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).cdc/tpcc-1000/rangefeed=true failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1845254&tab=buildLog) on [master@51e8960053a8224856234ffc1fb610c3843f7d7d](https://github.com/cockroachdb/cockroach/commits/51e8960053a8224856234ffc1fb610c3843f7d7d):

```
		       51.0s     2176            0.0            8.8      0.0      0.0      0.0      0.0 delivery
		       51.0s     2176            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		       51.0s     2176            0.0            8.7      0.0      0.0      0.0      0.0 orderStatus
		       51.0s     2176            0.0           20.7      0.0      0.0      0.0      0.0 payment
		       51.0s     2176            0.0            8.2      0.0      0.0      0.0      0.0 stockLevel
		       52.0s     2285            0.0            8.7      0.0      0.0      0.0      0.0 delivery
		       52.0s     2285            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		       52.0s     2285            0.0            8.5      0.0      0.0      0.0      0.0 orderStatus
		       52.0s     2285           47.0           21.2  49392.1  49392.1  49392.1  49392.1 payment
		       52.0s     2285            0.0            8.1      0.0      0.0      0.0      0.0 stockLevel:
		  - context canceled

	cluster.go:2368,cdc.go:224,cdc.go:521,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.cdcBasicTest
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:224
		    main.registerCDC.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:521
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 3: dead

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:825: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1845254-1585812734-27-n4cpu16 --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		1: 4385
		3: dead
		2: 4109
		Error: UNCLASSIFIED_PROBLEM:
		  - 3: dead
		    main.glob..func13
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1129
		    main.wrap.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:272
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:766
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:852
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:800
		    main.main
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1793
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:203
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/cdc/tpcc-1000/rangefeed=true](https://teamcity.cockroachdb.com/viewLog.html?buildId=1845254&tab=artifacts#/cdc/tpcc-1000/rangefeed=true)
Related:
- #46779 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #46355 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202003200044_v20.1.0-beta.3](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202003200044_v20.1.0-beta.3) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #46137 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202003112118_v20.1.0-beta.3](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202003112118_v20.1.0-beta.3) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #45437 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44971 roachtest: cdc/tpcc-1000/rangefeed=true failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Acdc%2Ftpcc-1000%2Frangefeed%3Dtrue.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
tac ~/Downloads/cockroach.log | grep -m 1 -B 30 -E '^[IWEF]' | tac
E200402 09:08:40.171928 343 util/log/crash_reporting.go:208  [n3,s3] a panic has occurred!
panic: lockTable bug - lock timestamp regression [recovered]
	panic: lockTable bug - lock timestamp regression

goroutine 343 [running]:
panic(0x3caf100, 0x4ac4fd0)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc024798560 sp=0xc0247984d0 pc=0x79a432
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000698000, 0x4bb4920, 0xc0008d41e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0x11f fp=0xc0247985c0 sp=0xc024798560 pc=0x126cf6f
runtime.call32(0x0, 0x44ab540, 0xc024799fa0, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:539 +0x3b fp=0xc0247985f0 sp=0xc0247985c0 pc=0x7c8d8b
panic(0x3caf100, 0x4ac4fd0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2 fp=0xc024798680 sp=0xc0247985f0 pc=0x79a322
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockState).acquireLock(0xc046a18240, 0x3, 0xc039806370, 0x1601f58912ce7d5b, 0xc000000001, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:1190 +0x98a fp=0xc024798770 sp=0xc024798680 pc=0x187e2da
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableImpl).AcquireLock(0xc003565720, 0xc039806370, 0xc038ea2b60, 0xb, 0x10, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:1828 +0x279 fp=0xc0247989e8 sp=0xc024798770 pc=0x1881aa9
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).OnLockAcquired(0xc003fc65c0, 0x4bb4920, 0xc0391ceb10, 0xc039806340)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:307 +0x77 fp=0xc024798a50 sp=0xc0247989e8 pc=0x1879e17
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleReadWriteLocalEvalResult(0xc0071c1500, 0x4bb4920, 0xc0391ceb10, 0x0, 0x0, 0x0, 0x0, 0xc039805000, 0x31, 0x45, ...)
```

--- a/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
@@ -60,7 +60,7 @@ import (
 // handle-write-intent-error  req=<req-name> txn=<txn-name> key=<key>
 // handle-txn-push-error      req=<req-name> txn=<txn-name> key=<key>  TODO(nvanbenschoten): implement this
 //
-// on-lock-acquired  req=<req-name> key=<key> [seq=<seq>]
+// on-lock-acquired  req=<req-name> key=<key> [seq=<seq>] [dur=r|u]
 // on-lock-updated   req=<req-name> txn=<txn-name> key=<key> status=[committed|aborted|pending] [ts=<int>[,<int>]]
 // on-txn-updated    txn=<txn-name> status=[committed|aborted|pending] [ts=<int>[,<int>]]
 //
@@ -276,6 +276,11 @@ func TestConcurrencyManagerBasic(t *testing.T) {
 				}
 				seqNum := enginepb.TxnSeq(seq)
 
+				dur := lock.Unreplicated
+				if d.HasArg("dur") {
+					dur = scanLockDurability(t, d)
+				}
+
 				// Confirm that the request has a corresponding write request.
 				found := false
 				for _, ru := range guard.Req.Requests {
@@ -298,7 +303,7 @@ func TestConcurrencyManagerBasic(t *testing.T) {
 				mon.runSync("acquire lock", func(ctx context.Context) {
 					log.Eventf(ctx, "txn %s @ %s", txn.ID.Short(), key)
 					span := roachpb.Span{Key: roachpb.Key(key)}
-					up := roachpb.MakeLockUpdateWithDur(txnAcquire, span, lock.Unreplicated)
+					up := roachpb.MakeLockUpdateWithDur(txnAcquire, span, dur)
 					m.OnLockAcquired(ctx, &up)
 				})
 				return c.waitAndCollect(t, mon)

--- a/pkg/kv/kvserver/concurrency/datadriven_util_test.go
+++ b/pkg/kv/kvserver/concurrency/datadriven_util_test.go
@@ -15,6 +15,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -56,6 +57,20 @@ func scanTimestampWithName(t *testing.T, d *datadriven.TestData, name string) hl
 	}
 	ts.Logical = int32(tsL)
 	return ts
+}
+
+func scanLockDurability(t *testing.T, d *datadriven.TestData) lock.Durability {
+	var durS string
+	d.ScanArgs(t, "dur", &durS)
+	switch durS {
+	case "r":
+		return lock.Replicated
+	case "u":
+		return lock.Unreplicated
+	default:
+		d.Fatalf(t, "unknown lock durability: %s", durS)
+		return 0
+	}
 }
 
 func scanSingleRequest(

--- a/pkg/kv/kvserver/concurrency/lock_table.go
+++ b/pkg/kv/kvserver/concurrency/lock_table.go
@@ -1179,16 +1179,35 @@ func (l *lockState) acquireLock(
 		//  - txn A's coordinator does not immediately learn of the push
 		//  - txn A re-acquires lock at sequence 2, ts 15
 		//
-		// A lock's timestamp cannot be allowed to regress, so by forwarding
-		// its timestamp during the second acquisition instead if assigning
-		// to it blindly, it remains at 20.
+		// A lock's timestamp at a given durability level is not allowed to
+		// regress, so by forwarding its timestamp during the second acquisition
+		// instead if assigning to it blindly, it remains at 20.
+		//
+		// However, a lock's timestamp as reported by getLockerInfo can regress
+		// if it is acquired at a lower timestamp and a different durability
+		// than it was previously held with. This is necessary to support
+		// because the hard constraint which we must uphold here that the
+		// lockHolderInfo for a replicated lock cannot diverge from the
+		// replicated state machine in such a way that its timestamp in the
+		// lockTable exceeds that in the replicated keyspace. If this invariant
+		// were to be violated, we'd risk infinite lock-discovery loops for
+		// requests that conflict with the lock as is written in the replicated
+		// state machine but not as is reflected in the lockTable.
+		//
+		// Lock timestamp regressions are safe from the perspective of other
+		// transactions because the request which re-acquired the lock at the
+		// lower timestamp must have been holding a write latch at or below the
+		// new lock's timestamp. This means that no conflicting requests could
+		// be evaluating concurrently. Instead, all will need to re-scan the
+		// lockTable once they acquire latches and will notice the reduced
+		// timestamp at that point, which may cause them to conflict with the
+		// lock even if they had not conflicted before. In a sense, it is no
+		// different than the first time a lock is added to the lockTable.
 		l.holder.holder[durability].ts.Forward(ts)
 		l.holder.holder[durability].seqs = append(seqs, txn.Sequence)
 
 		_, afterTs, _ := l.getLockerInfo()
-		if afterTs.Less(beforeTs) {
-			panic("lockTable bug - lock timestamp regression")
-		} else if beforeTs.Less(afterTs) {
+		if beforeTs.Less(afterTs) {
 			l.increasedLockTs(afterTs)
 		}
 		return nil

--- a/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/update
+++ b/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/update
@@ -5,7 +5,7 @@
 # still at the original timestamp. This is permitted but the
 # lock's timestamp should not regress.
 #
-# Setup: txn1 acquire lock k
+# Setup: txn1 acquires lock k
 #        txn2 reads k and waits
 #        txn2 pushes txn1
 #
@@ -124,7 +124,7 @@ reset namespace
 # the original timestamp. This is permitted but the lock's
 # timestamp should not regress.
 #
-# Setup: txn1 acquire lock k
+# Setup: txn1 acquires lock k
 #        txn2 reads k and waits
 #
 # Test:  txn2 pushes txn1's timestamp forward
@@ -234,6 +234,130 @@ debug-lock-table
 global: num=1
  lock: "k"
   holder: txn: 00000001-0000-0000-0000-000000000000, ts: 0.000000012,2, info: unrepl epoch: 1, seqs: [0]
+local: num=0
+
+reset namespace
+----
+
+# -------------------------------------------------------------
+# A transaction acquires an unreplicated lock. The lock is
+# pushed to a higher timestamp by a second transaction. The
+# transaction then returns to upgrade the unreplicated lock to a
+# replicated intent at a new sequence number but still at the
+# original timestamp. This is permitted and the lock's timestamp
+# regresses back down to the intent's timestamp. In practice, if
+# the pusher wanted to prevent its push from being reverted, it
+# should have also bumped the timestamp cache to ensure that the
+# intent couldn't be laid down at the original timestamp.
+#
+# Setup: txn1 acquires unreplicated lock k
+#        txn2 reads k and waits
+#        txn2 pushes txn1
+#
+# Test:  txn2 succeeds in pushing txn1's ts forward
+#        txn2 proceeds
+#        txn1 re-acquires replicated lock k at lower ts
+# -------------------------------------------------------------
+
+new-txn name=txn1 ts=10,1 epoch=0
+----
+
+new-txn name=txn2 ts=12,1 epoch=0
+----
+
+new-request name=req1 txn=txn1 ts=10,1
+  put key=k value=v
+----
+
+new-request name=req2 txn=txn2 ts=12,1
+  get key=k
+----
+
+sequence req=req1
+----
+[1] sequence req1: sequencing request
+[1] sequence req1: acquiring latches
+[1] sequence req1: scanning lock table for conflicting locks
+[1] sequence req1: sequencing complete, returned guard
+
+on-lock-acquired req=req1 key=k dur=u
+----
+[-] acquire lock: txn 00000001 @ k
+
+finish req=req1
+----
+[-] finish req1: finishing request
+
+sequence req=req2
+----
+[2] sequence req2: sequencing request
+[2] sequence req2: acquiring latches
+[2] sequence req2: scanning lock table for conflicting locks
+[2] sequence req2: waiting in lock wait-queues
+[2] sequence req2: pushing timestamp of txn 00000001 above 0.000000012,1
+[2] sequence req2: blocked on select in concurrency_test.(*cluster).PushTransaction
+
+debug-lock-table
+----
+global: num=1
+ lock: "k"
+  holder: txn: 00000001-0000-0000-0000-000000000000, ts: 0.000000010,1, info: unrepl epoch: 0, seqs: [0]
+   waiting readers:
+    req: 8, txn: 00000002-0000-0000-0000-000000000000
+   distinguished req: 8
+local: num=0
+
+# --------------------------------
+# Setup complete, test starts here
+# --------------------------------
+
+on-txn-updated txn=txn1 status=pending ts=12,2
+----
+[-] update txn: increasing timestamp of txn1
+[2] sequence req2: resolving intent "k" for txn 00000001 with PENDING status
+[2] sequence req2: acquiring latches
+[2] sequence req2: scanning lock table for conflicting locks
+[2] sequence req2: sequencing complete, returned guard
+
+finish req=req2
+----
+[-] finish req2: finishing request
+
+debug-lock-table
+----
+global: num=1
+ lock: "k"
+  holder: txn: 00000001-0000-0000-0000-000000000000, ts: 0.000000012,2, info: unrepl epoch: 0, seqs: [0]
+local: num=0
+
+# Issue another write to the same key for txn1 at its initial timestamp,
+# this time with a replicated durability. The timestamp in the lock
+# table should regress back down to reflect the replicated lock state.
+
+new-request name=req3 txn=txn1 ts=10,1
+  put key=k value=v2 seq=1
+----
+
+sequence req=req3
+----
+[3] sequence req3: sequencing request
+[3] sequence req3: acquiring latches
+[3] sequence req3: scanning lock table for conflicting locks
+[3] sequence req3: sequencing complete, returned guard
+
+on-lock-acquired req=req3 key=k seq=1 dur=r
+----
+[-] acquire lock: txn 00000001 @ k
+
+finish req=req3
+----
+[-] finish req3: finishing request
+
+debug-lock-table
+----
+global: num=1
+ lock: "k"
+  holder: txn: 00000001-0000-0000-0000-000000000000, ts: 0.000000010,1, info: repl epoch: 0, seqs: [1], unrepl epoch: 0, seqs: [0]
 local: num=0
 
 reset namespace

--- a/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_changes
+++ b/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_changes
@@ -50,6 +50,19 @@ global: num=1
 local: num=0
 
 # ---------------------------------------------------------------------------------
+# Lock is reacquired at same epoch with lower timestamp and different durability.
+# The lock's timestamp is allowed to regress in this case, because it must never
+# diverge from the replicated state machine.
+# ---------------------------------------------------------------------------------
+
+acquire r=req2 k=a durability=r
+----
+global: num=1
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000008,0, info: repl epoch: 0, seqs: [3], unrepl epoch: 0, seqs: [2, 3]
+local: num=0
+
+# ---------------------------------------------------------------------------------
 # Lock is reacquired at a different epoch. The old sequence numbers are discarded.
 # ---------------------------------------------------------------------------------
 
@@ -63,7 +76,7 @@ acquire r=req3 k=a durability=u
 ----
 global: num=1
  lock: "a"
-  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000010,0, info: unrepl epoch: 1, seqs: [0]
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000008,0, info: repl epoch: 0, seqs: [3], unrepl epoch: 1, seqs: [0]
 local: num=0
 
 # ---------------------------------------------------------------------------------
@@ -72,17 +85,17 @@ local: num=0
 # discarded but the timestamp is not regressed.
 # ---------------------------------------------------------------------------------
 
-new-txn txn=txn1 ts=8 epoch=2 seq=0
+new-txn txn=txn1 ts=6 epoch=2 seq=0
 ----
 
-new-request r=req4 txn=txn1 ts=8 spans=w@a
+new-request r=req4 txn=txn1 ts=6 spans=w@a
 ----
 
 acquire r=req4 k=a durability=u
 ----
 global: num=1
  lock: "a"
-  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000010,0, info: unrepl epoch: 2, seqs: [0]
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000008,0, info: repl epoch: 0, seqs: [3], unrepl epoch: 2, seqs: [0]
 local: num=0
 
 # ---------------------------------------------------------------------------------
@@ -98,13 +111,13 @@ start-waiting: true
 
 guard-state r=req5
 ----
-new: state=waitForDistinguished txn=txn1 ts=10 key="a" held=true guard-access=read
+new: state=waitForDistinguished txn=txn1 ts=8 key="a" held=true guard-access=read
 
 print
 ----
 global: num=1
  lock: "a"
-  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000010,0, info: unrepl epoch: 2, seqs: [0]
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000008,0, info: repl epoch: 0, seqs: [3], unrepl epoch: 2, seqs: [0]
    waiting readers:
     req: 1, txn: 00000000-0000-0000-0000-000000000002
    distinguished req: 1
@@ -116,11 +129,25 @@ new-txn txn=txn1 ts=14 epoch=1 seq=1
 new-request r=req6 txn=txn1 ts=14 spans=w@a
 ----
 
+acquire r=req6 k=a durability=r
+----
+global: num=1
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000010,0, info: repl epoch: 1, seqs: [1], unrepl epoch: 2, seqs: [0]
+   waiting readers:
+    req: 1, txn: 00000000-0000-0000-0000-000000000002
+   distinguished req: 1
+local: num=0
+
+guard-state r=req5
+----
+old: state=waitForDistinguished txn=txn1 ts=8 key="a" held=true guard-access=read
+
 acquire r=req6 k=a durability=u
 ----
 global: num=1
  lock: "a"
-  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000014,0, info: unrepl epoch: 1, seqs: [0, 1]
+  holder: txn: 00000000-0000-0000-0000-000000000001, ts: 0.000000014,0, info: repl epoch: 1, seqs: [1], unrepl epoch: 1, seqs: [0, 1]
 local: num=0
 
 guard-state r=req5

